internal/types/subtypes/attribute_transform.go

```diff
@@ -112,3 +112,57 @@ func protoToStruct(p proto.Message) (*structpb.Struct, error) {
 	}
 	return st, nil
 }
+
+// Filterable converts a proto.Message so any subtype attributes fields are
+// structed like the API so filter strings will be correctly applied. If the
+// given proto.Message does not have any subtype attributes, the original
+// proto.Message is returned.
+func Filterable(item proto.Message) (proto.Message, error) {
+	clone := proto.Clone(item)
+	r := clone.ProtoReflect()
+
+	attrsField := r.Descriptor().Oneofs().ByName("attrs")
+	if attrsField == nil {
+		return item, nil
+	}
+
+	defaultAttrField := attrsField.Fields().ByName("attributes")
+	if defaultAttrField == nil {
+		return item, nil
+	}
+
+	var attr proto.Message
+	var pbAttrs proto.Message
+	var err error
+
+	oneOfFields := attrsField.Fields()
+
+	for i := 0; i < oneOfFields.Len(); i++ {
+		attrField := oneOfFields.Get(i)
+		attr = r.Get(attrField).Message().Interface()
+		if attrField == defaultAttrField {
+			if attr != (*structpb.Struct)(nil) {
+				pbAttrs = attr
+				break
+			}
+		} else if attr != nil {
+			pbAttrs, err = protoToStruct(attr)
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+	}
+
+	// no attrs set, so the original item can just be filtered as is
+	if pbAttrs == nil {
+		return item, nil
+	}
+
+	r.Set(defaultAttrField, protoreflect.ValueOfMessage(pbAttrs.ProtoReflect()))
```

Comment:
Worth commenting that this also unset the old field, by virtue of `defaultAttrField` being part of a `oneof`.

Reply:
That is documents on [Set](https://github.com/protocolbuffers/protobuf-go/blob/32051b4f86e54c2142c7c05362c6e96ae3454a1c/reflect/protoreflect/value.go#L95-L96) do you think it is needed here as well? We call set on a one of in a couple spots in this file. Would you want a comment in both places?

Reply:
I think it would be nice, yeah.